Kml Tour Documentation

srothst1 opened the pull request:
In response to issue #8073

Updated documentation for `KmlTourWait`, `KmlTourFlyTo`, and `KmlTour`. Let me know if there are any suggested edits.

The current documentation references `KmlTourSoundCues` various times in `KmlTour`. This is seen here:
[https://cesium.com/docs/cesiumjs-ref-doc/KmlTour.html?classFilter=KmlTour](https://cesium.com/docs/cesiumjs-ref-doc/KmlTour.html?classFilter=KmlTour)

I was not able to find `KmlTourSoundCues` anywhere else in the documentation or in the code - how should we handle this?

@shehzan10
@OmarShehata

cesium-concierge commented:
Thanks for the pull request @srothst1!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/KML-tour-doc/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

ebogo1 commented:
These changes look good to me. 

1. If you use the syntax "Fixes #XXXX" in the first post then this PR should get linked to the issue and close it automatically when merged.
2. Re: `KmlTourSoundCues` - I think it might be fine to remove this from the docs.. there's no other instance of `KmlTourSoundCues` in the codebase as far as I can tell, and it looks like it gets processed with `processTourUnsupportedNode()`. @lilleyse Is there a plan for adding more node types here?

srothst1 commented:
@ebogo1 thank you for the feedback! I agree with your thoughts on `KmlTourSoundCues`. However, I am interested to see what @lilleyse thinks.

lilleyse commented on Source/DataSources/KmlTour.js:
```diff
@@ -1,12 +1,20 @@
+ * @see KmlTourFlyTo
+ * @see KmlTourWait
+ *
+ * @demo {@link https://sandcastle.cesium.com/?src=KML%20Tours.html KML Tours}
```
The generated doc looked ok but generally there's a `|` symbol between the link and name

```suggestion
 * @demo {@link https://sandcastle.cesium.com/?src=KML%20Tours.html|KML Tours}
```

lilleyse commented:
This PR looks mostly good but it's missing a important piece from the original issue: the `kmlTours` property is used in the sandcastle but is missing from the `KmlDataSource` documentation.

lilleyse commented:
I agree about removing `KmlTourSoundCues` from the docs.

> @lilleyse Is there a plan for adding more node types here?

If there's not an issue for this already we can open one, though it's probably low-ish priority.

srothst1 commented:
@lilleyse I just removed all references to `KmlTourSoundCues` in `KmlTour`. I also added the `kmlTours` property to the `KmlDataSource` documentation. Please let me know if you have any feedback.

ebogo1 commented:
> If there's not an issue for this already we can open one, though it's probably low-ish priority.

opened #9602

lilleyse commented on Source/DataSources/KmlDataSource.js:
```diff
@@ -3349,6 +3349,7 @@ function load(dataSource, entityCollection, data, options) {
  * @property {Boolean} [clampToGround=false] true if we want the geometry features (Polygons, LineStrings and LinearRings) clamped to the ground.
  * @property {Ellipsoid} [ellipsoid=Ellipsoid.WGS84] The global ellipsoid used for geographical calculations.
  * @property {Credit|String} [credit] A credit for the data source, which is displayed on the canvas.
+ * @property {KmlTour} [KmlTours] The KmlTour that is used to guide the camera to specified destinations on given time intervals.
```
This isn't the correct place to put the documentation since `kmlTours` is not a property of the constructor options but rather a property of the data source itself. The fix will require a few changes:

* Add a private member variable to `KmlDataSource`. In the constructor right underneath `this._resourceCredits` add `this._kmlTours = [];`
* Add a `kmlTours` getter to `Object.defineProperties`. No need to add a setter. It will will look something like:

```js
  /**
   * Gets the KML Tours that are used to guide the camera to specified destinations on given time intervals.
   * @memberof KmlDataSource.prototype
   * @type {KmlTour[]}
   */
  credit: {
    get: function () {
      return this._credit;
    },
  }
```
* Any code inside `KmlDataSource` should use `_kmlTours`; anything outside should use `kmlTours`. For example, the Sandcastle will continue to use `kmlTours`.
* The following lines can be removed because the empty array will be initialized in the constructor
```js
if (!defined(dataSource.kmlTours)) {
    dataSource.kmlTours = [];
}
```

srothst1 replied:
@lilleyse thank you for the feedback. I just added a new commit that should address your suggestions.

lilleyse commented:
@srothst1 perfect! The generated documentation looks good, unit tests pass locally, and the KML sandcastle works as expected.

One last thing, since `kmlTours` is now part of the public API could you mention it in the additions section of `CHANGES.md`?
